### PR TITLE
feat: Integration for Gro Protocol

### PR DESCRIPTION
**Pull request by farhaan-ali**
## Description

Adding Gro Protocol under Apps:
- Ability for users to view their labs balances on AVAX
- Ability for users to view their pool balances on ETH

## Checklist

- [X] I have followed the [Contributing Guidelines](../CONTRIBUTING.md)

## How to test?

address to test ETH pool balances: 0xbd722F41bca276B05a27E0B716bdeC2cB801D952
address to test AVAX Labs balances: 0x9b4A2a6D9c244c8c94e31A84A169E23b4b7E5b1f

### CODE REVIEW

File: src/apps/gro/avalanche/gro.labs.token-fetcher.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    const tokens = await Promise.all(
+      vaultAddresses.map(async vaultAddress => {
+        const contract = this.groContractFactory.groLabsVault({ address: vaultAddress, network });
+
+        const [symbol, decimals, supplyRaw, pricePerShare, name] = await Promise.all([
+          multicall.wrap(contract).symbol(),
+          multicall.wrap(contract).decimals(),
+          multicall.wrap(contract).totalSupply(),
+          multicall.wrap(contract).getPricePerShare(),
+          multicall.wrap(contract).name(),
```
Comment (immasandwich) [inline src/apps/gro/avalanche/gro.labs.token-fetcher.ts:42]: Your `tokens` array below is empty, but it should be the underlying token of the lab vault token.

1. Above, under line 31, get the base tokens.

```ts
const baseTokens = await this.appToolkit.getBaseTokenPrices(network);
```

2. Here, add a call to `token` to get the underlying token in the lab vault.

```ts
const [symbol, decimals, supplyRaw, pricePerShare, name, underlyingTokenAddressRaw] = await Promise.all([
  multicall.wrap(contract).symbol(),
  multicall.wrap(contract).decimals(),
  multicall.wrap(contract).totalSupply(),
  multicall.wrap(contract).getPricePerShare(),
  multicall.wrap(contract).name(),
  multicall.wrap(contract).token(), 
]);
```

3. Match the underlying token, and put it in `tokens`.

```ts
const underlyingToken = baseTokens.find(v => v.address === underlyingTokenAddressRaw.toLowerCase();
const tokens = [underlyingToken];
```

File: src/apps/gro/avalanche/gro.labs.token-fetcher.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          supply,
+          pricePerShare: Number(pricePerShare),
+          price: Number(pricePerShare),
+          tokens,
+          dataProps: {},
+          displayProps: {
+            label: name,
+            images: getImagesFromToken(underlyingToken!),
+            secondaryLabel: buildDollarDisplayItem(Number(pricePerShare)),
+            tertiaryLabel: undefined,
```
Comment (immasandwich) [inline src/apps/gro/avalanche/gro.labs.token-fetcher.ts:72]: You can just remove the line `tertiaryLabel: undefined`.

File: src/apps/gro/avalanche/gro.pools.contract-position-fetcher.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+@Register.ContractPositionFetcher({ appId, groupId, network })
+export class AvalancheGroPoolsContractPositionFetcher implements PositionFetcher<ContractPosition> {
+  constructor(
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(GroContractFactory) private readonly groContractFactory: GroContractFactory,
+  ) {}
+
+  async getPositions() {
+    return [];
```
Comment (immasandwich) [inline src/apps/gro/avalanche/gro.pools.contract-position-fetcher.ts:24]: Looks like we don't need this file?

File: src/apps/gro/avalanche/gro.vesting.contract-position-fetcher.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+import { GroContractFactory } from '../contracts';
+import { GRO_DEFINITION } from '../gro.definition';
+
+const appId = GRO_DEFINITION.id;
+const groupId = GRO_DEFINITION.groups.vesting.id;
+const network = Network.AVALANCHE_MAINNET;
+
+@Register.ContractPositionFetcher({ appId, groupId, network })
+export class AvalancheGroVestingContractPositionFetcher implements PositionFetcher<ContractPosition> {
```
Comment (immasandwich) [inline src/apps/gro/avalanche/gro.vesting.contract-position-fetcher.ts:17]: And we can delete this as well. Probably autogenerated.

File: src/apps/gro/ethereum/gro.labs.token-fetcher.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+@Register.TokenPositionFetcher({ appId, groupId, network })
+export class EthereumGroLabsTokenFetcher implements PositionFetcher<AppTokenPosition> {
+  constructor(
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(GroContractFactory) private readonly groContractFactory: GroContractFactory,
+  ) {}
+
+  async getPositions() {
+    return [];
```
Comment (immasandwich) [inline src/apps/gro/ethereum/gro.labs.token-fetcher.ts:24]: Will this be implemented?

File: src/apps/gro/ethereum/gro.pools.contract-position-fetcher.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+@Register.ContractPositionFetcher({ appId, groupId, network })
+export class EthereumGroPoolsContractPositionFetcher implements PositionFetcher<ContractPosition> {
+  constructor(
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(GroContractFactory) private readonly groContractFactory: GroContractFactory,
+  ) {}
+
+  async getPositions(): Promise<ContractPosition<DefaultDataProps>[]> {
+    return this.appToolkit.helpers.masterChefContractPositionHelper.getContractPositions<GroLpTokenStaker>({
+      address: '0x2E32bAd45a1C29c1EA27cf4dD588DF9e68ED376C',
```
Comment (immasandwich) [inline src/apps/gro/ethereum/gro.pools.contract-position-fetcher.ts:27]: Lowercase.

File: src/apps/gro/ethereum/gro.pools.contract-position-fetcher.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+      network,
+      dependencies: [
+        {
+          appId: CURVE_DEFINITION.id,
+          groupIds: [CURVE_DEFINITION.groups.pool.id],
+          network,
+        },
+        {
+          appId: 'uniswap-v2',
+          groupIds: ['pool'],
```
Comment (immasandwich) [inline src/apps/gro/ethereum/gro.pools.contract-position-fetcher.ts:39]: Import `UNISWAP_V2_DEFINITION` for these values.

**Comment by farhaan-ali** [comment]
Example of an address that should have vesting position balances returned 0x6523f418c4b7260b070a17c47a206b66935d67cd

File: src/apps/gro/ethereum/gro.vesting.contract-position-fetcher.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(GroContractFactory) private readonly groContractFactory: GroContractFactory,
+  ) {}
+
+  async getPositions() {
+    const groVestingAddress = '0x748218256AfE0A19a88EBEB2E0C5Ce86d2178360';
+    const groDaoToken = '0x3Ec8798B81485A254928B70CDA1cf0A2BB0B74D7';
+    const baseTokens = await this.appToolkit.getBaseTokenPrices(network);
+    const underlyingToken = baseTokens.find(v => v.address === groDaoToken.toLowerCase());
+    const tokens = [vesting(underlyingToken!)];
```
Comment (immasandwich) [inline src/apps/gro/ethereum/gro.vesting.contract-position-fetcher.ts:30]: You likely want `const tokens = [locked(underlyingToken!), claimable(underlyingToken!)]`

The `locked` token represents the amount that is locked, and the `claimable` token represents the amount that is unlocked.

File: src/apps/gro/ethereum/gro.vesting.contract-position-fetcher.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+@Register.ContractPositionFetcher({ appId, groupId, network })
+export class EthereumGroVestingContractPositionFetcher implements PositionFetcher<ContractPosition> {
+  constructor(
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(GroContractFactory) private readonly groContractFactory: GroContractFactory,
+  ) {}
+
+  async getPositions() {
+    const groVestingAddress = '0x748218256AfE0A19a88EBEB2E0C5Ce86d2178360';
+    const groDaoToken = '0x3Ec8798B81485A254928B70CDA1cf0A2BB0B74D7';
```
Comment (immasandwich) [inline src/apps/gro/ethereum/gro.vesting.contract-position-fetcher.ts:27]: Lowercase all addresses.

File: src/apps/gro/ethereum/gro.vesting.contract-position-fetcher.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+    const position: ContractPosition = {
+      type: ContractType.POSITION,
+      appId,
+      groupId,
+      address: groVestingAddress,
+      network,
+      tokens,
+      dataProps: {},
+      displayProps: { label: '', images: [] },
```
Comment (immasandwich) [inline src/apps/gro/ethereum/gro.vesting.contract-position-fetcher.ts:40]: `label` should be:

```
const label = `Vesting ${underlyingToken!.symbol}`;
```

File: src/apps/gro/ethereum/gro.balance-fetcher.ts [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  }
+
+  private async getVestingBalances(address: string) {
+    return this.appToolkit.helpers.contractPositionBalanceHelper.getContractPositionBalances({
+      address,
+      appId: GRO_DEFINITION.id,
+      groupId: GRO_DEFINITION.groups.pools.id,
+      network,
+      resolveBalances: async ({ address, contractPosition, multicall }) => {
+        const groVestingAddress = '0x748218256AfE0A19a88EBEB2E0C5Ce86d2178360';
```
Comment (immasandwich) [inline src/apps/gro/ethereum/gro.balance-fetcher.ts:55]: Lowecase.

File: src/apps/gro/ethereum/gro.balance-fetcher.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        const contract = this.groContractFactory.groVesting({ network, address: groVestingAddress });
+        const vestedToken = contractPosition.tokens.find(isVesting)!;
```
Comment (immasandwich) [inline src/apps/gro/ethereum/gro.balance-fetcher.ts:57]: As mentioned below, you'll use the `locked` and `claimable` metatypes, so:

```ts
const lockedToken = contractPosition.tokens.find(isLocked)!;
const unlockedToken = contractPosition.tokens.find(isClaimable)!;
```

File: src/apps/gro/ethereum/gro.balance-fetcher.ts [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+            .then(v => v.rewardDebt),
+      }),
+    });
+  }
+
+  private async getVestingBalances(address: string) {
+    return this.appToolkit.helpers.contractPositionBalanceHelper.getContractPositionBalances({
+      address,
+      appId: GRO_DEFINITION.id,
+      groupId: GRO_DEFINITION.groups.farm.id,
```
Comment (immasandwich) [inline src/apps/gro/ethereum/gro.balance-fetcher.ts:52]: Should be: 

```ts
groupId: GRO_DEFINITION.groups.vesting.id,
```

**Comment by SamIam-0x** [comment]
Hello @0xfarhaan !

Looks like you did not provide an address when submitting your PR. I recognize it may have seemed kind of random that we were asking for a wallet address with a PR submission, but we are actually looking for ways to reward Studio contributors in the future. 

Is there an address you’d want to provide? Could even be an empty address, if you do not have an doxxed address

Let me know! Also, if you’d like to privately provide it, feel free to shoot me a TG https://t.me/SamIAm_0x or DM on the Zapper Discord.

Thanks!
Sam

**Pull request merged**